presto-main/src/main/java/io/prestosql/operator/StreamingAggregationOperator.java

```diff
@@ -232,17 +230,25 @@ private StreamingAggregation(
                 if (outputPages.isEmpty()) {
                     return finished();
                 }
+
+                return ofResult(outputPages.removeFirst(), false);
             }
-            else {
-                processInput(inputPage);
-                updateMemoryUsage();
+
+            // flush remaining output pages before requesting next input page
+            if (!outputPages.isEmpty()) {
+                Page outputPage = outputPages.removeFirst();
+                return ofResult(outputPage, outputPages.isEmpty());
             }
 
+            processInput(inputPage);
+            updateMemoryUsage();
+
             if (outputPages.isEmpty()) {
                 return needsMoreData();
             }
 
-            return ofResult(outputPages.removeFirst(), !finishing);
+            Page outputPage = outputPages.removeFirst();
+            return ofResult(outputPage, outputPages.isEmpty());
```

Comment:
this should be 

```suggestion
            return ofResult(outputPage, true);
```

because `inputPage` is fully consumed and should not be re-submitted

Reply:
I've added comment about this invariant